crates/proof-of-sql/src/base/database/join_util.rs

```diff
@@ -1,19 +1,93 @@
 use super::{
     apply_column_to_indexes,
     order_by_util::{compare_indexes_by_columns, compare_single_row_of_tables},
-    Column, ColumnRepeatOp, ElementwiseRepeatOp, RepetitionOp, Table, TableOperationError,
-    TableOperationResult, TableOptions,
+    union_util::column_union,
+    Column, ColumnOperationResult, ColumnRepeatOp, ElementwiseRepeatOp, RepetitionOp, Table,
+    TableOperationError, TableOperationResult, TableOptions,
 };
 use crate::base::{
     map::{IndexMap, IndexSet},
     scalar::Scalar,
 };
-use alloc::vec::Vec;
+use alloc::{vec, vec::Vec};
 use bumpalo::Bump;
 use core::cmp::Ordering;
 use itertools::Itertools;
 use sqlparser::ast::Ident;
 
+/// Compute the set union of two slices of columns, deduplicate and sort the result.
+///
+/// Notes
+/// 1. This is mostly used for joins.
+/// 2. We do not check whether columns in the args have the same length, as we assume that the columns in an arg are already from the same table.
+pub(crate) fn ordered_set_union<'a, S: Scalar>(
+    left_on: &[Column<'a, S>],
+    right_on: &[Column<'a, S>],
+    alloc: &'a Bump,
+) -> TableOperationResult<Vec<Column<'a, S>>> {
+    //1. Union the columns
+    if left_on.len() != right_on.len() {
+        return Err(TableOperationError::JoinWithDifferentNumberOfColumns {
+            left_num_columns: left_on.len(),
+            right_num_columns: right_on.len(),
+        });
+    }
```

Comment:
I think it's reasonable to assert here instead. Users should never be able to hit this code.